### DISCUSSION TITLE
zh/integrations/coreml/

### DISCUSSION THREAD

**Discussion by giscus** [ultralytics/ultralytics#9766] [Docs]
# zh/integrations/coreml/

探索将Ultralytics YOLOv8  模型导出为CoreML 格式的过程，为苹果设备上的 iOS 和 macOS 应用程序提供高效的对象检测功能。

https://docs.ultralytics.com/zh/integrations/coreml/

**Comment by ChristalZhu** [comment]
我按照https://harshavardhanyellanki.github.io/YOLOV8_GUI/tasks/segmentation/的方法，将我自己训练的分割模型进行导出成mlpackage，但是出现了问题就是打开这个mlpackage之后，缺少class label文件和缺少了preview，请问这个有什么办法解决，如果可以的话，麻烦发送到我的email账号：wenpeng.zhu@foxmail.com万分感谢。
I followed the method at https://harshavardhanyellanki.github.io/YOLOV8_GUI/tasks/segmentation/ to export the segmentation model I trained myself into mlpackage, but the problem is that after opening this mlpackage, the class label file and the preview are missing. Is there any way to solve this problem? If possible, please send it to my email account: wenpeng.zhu@foxmail.com. Thank you very much.

**Reply by pderrenger** [reply]
Hello @ChristalZhu,

Thank you for reaching out and providing details about the issue you're facing with exporting your YOLOv8 segmentation model to CoreML. Let's work together to resolve this.

Firstly, please ensure that you are using the latest versions of `torch` and `ultralytics`. You can update them using the following commands:
```bash
pip install --upgrade torch ultralytics
```

If the issue persists, could you please provide a minimum reproducible code example? This will help us better understand the problem and investigate a solution. You can refer to our guide on creating a minimum reproducible example here: [Minimum Reproducible Example](https://docs.ultralytics.com/help/minimum_reproducible_example).

Regarding the missing class label file and preview in the `.mlpackage`, this can sometimes occur if the export process does not include all necessary metadata. Here is a Python example to ensure the export includes all required components:

```python
from ultralytics import YOLO

# Load your trained YOLOv8 model
model = YOLO("path/to/your/custom_model.pt")

# Export the model to CoreML format
model.export(format="coreml", nms=True)  # Ensure NMS is included for proper detection
```

After exporting, you can manually check the `.mlpackage` to ensure it contains the necessary class labels and preview files. If they are still missing, please let us know, and we can further investigate.

For more detailed instructions on exporting to CoreML, you can visit our documentation here: [CoreML Export Guide](https://docs.ultralytics.com/integrations/coreml/).

Feel free to share any additional details or errors you encounter. We're here to help!

**Reply by ChristalZhu** [reply]
Still the same problem, this is the result of running in the terminal：
PyTorch: starting from '/Users/doublepeng/Desktop/best.pt' with input shape (1, 3, 640, 640) BCHW and output shape(s) ((1, 37, 8400), (1, 32, 160, 160)) (6.5 MB)

CoreML: starting export with coremltools 7.2...
CoreML: WARNING ⚠️ 'nms=True' is only available for Detect models like 'yolov8n.pt'.
Tuple detected at graph output. This will be flattened in the converted model.
Converting PyTorch Frontend ==> MIL Ops: 100%|███████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████▌| 679/681 [00:00<00:00, 8273.25 ops/s]
Running MIL frontend_pytorch pipeline: 100%|████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████| 5/5 [00:00<00:00, 157.94 passes/s]
Running MIL default pipeline: 100%|████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████| 78/78 [00:00<00:00, 82.56 passes/s]
Running MIL backend_mlprogram pipeline: 100%|█████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████| 12/12 [00:00<00:00, 160.71 passes/s]
CoreML: export success ✅ 7.7s, saved as '/Users/doublepeng/Desktop/best.mlpackage' (6.4 MB)

Export complete (8.1s)
Results saved to /Users/doublepeng/Desktop
Predict:         yolo predict task=segment model=/Users/doublepeng/Desktop/best.mlpackage imgsz=640  
Validate:        yolo val task=segment model=/Users/doublepeng/Desktop/best.mlpackage imgsz=640 data=/home/yujialin/Documents/xiaotong/corn/data/corn.yaml  
Visualize:       https://netron.app
In the run mlpackage file, the class label is still missing in General, and there is no preview. Since I am only doing corn segmentation, if possible, can I send you my .pt file?

**Reply by pderrenger** [reply]
Hello @ChristalZhu,

Thank you for providing the detailed output from your terminal. It looks like the CoreML export process completed successfully, but you're encountering issues with missing class labels and preview in the `.mlpackage`.

To address this, please ensure you are using the latest versions of `torch` and `ultralytics`. You can update them using the following commands:
```bash
pip install --upgrade torch ultralytics
```

If the issue persists, could you please provide a minimum reproducible code example? This will help us better understand the problem and investigate a solution. You can refer to our guide on creating a minimum reproducible example here: [Minimum Reproducible Example](https://docs.ultralytics.com/help/minimum_reproducible_example).

Regarding the missing class label file and preview in the `.mlpackage`, this can sometimes occur if the export process does not include all necessary metadata. Here is a Python example to ensure the export includes all required components:

```python
from ultralytics import YOLO

# Load your trained YOLOv8 model
model = YOLO("path/to/your/custom_model.pt")

# Export the model to CoreML format
model.export(format="coreml", nms=True)  # Ensure NMS is included for proper detection
```

After exporting, you can manually check the `.mlpackage` to ensure it contains the necessary class labels and preview files. If they are still missing, please let us know, and we can further investigate.

For more detailed instructions on exporting to CoreML, you can visit our documentation here: [CoreML Export Guide](https://docs.ultralytics.com/integrations/coreml/).

Feel free to share any additional details or errors you encounter. We're here to help! 😊

**Comment by ChristalZhu** [comment]
Still the same problem, this is the result of running in the terminal：
PyTorch: starting from '/Users/doublepeng/Desktop/best.pt' with input shape (1, 3, 640, 640) BCHW and output shape(s) ((1, 37, 8400), (1, 32, 160, 160)) (6.5 MB)

CoreML: starting export with coremltools 7.2...
CoreML: WARNING ⚠️ 'nms=True' is only available for Detect models like 'yolov8n.pt'.
Tuple detected at graph output. This will be flattened in the converted model.
Converting PyTorch Frontend ==> MIL Ops: 100%|███████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████▌| 679/681 [00:00<00:00, 8273.25 ops/s]
Running MIL frontend_pytorch pipeline: 100%|████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████| 5/5 [00:00<00:00, 157.94 passes/s]
Running MIL default pipeline: 100%|████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████| 78/78 [00:00<00:00, 82.56 passes/s]
Running MIL backend_mlprogram pipeline: 100%|█████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████| 12/12 [00:00<00:00, 160.71 passes/s]
CoreML: export success ✅ 7.7s, saved as '/Users/doublepeng/Desktop/best.mlpackage' (6.4 MB)

Export complete (8.1s)
Results saved to /Users/doublepeng/Desktop
Predict: yolo predict task=segment model=/Users/doublepeng/Desktop/best.mlpackage imgsz=640
Validate: yolo val task=segment model=/Users/doublepeng/Desktop/best.mlpackage imgsz=640 data=/home/yujialin/Documents/xiaotong/corn/data/corn.yaml
Visualize: https://netron.app
In the run mlpackage file, the class label is still missing in General, and there is no preview. Since I am only doing corn segmentation, if possible, can I send you my .pt file?

**Reply by glenn-jocher** [reply]
@ChristalZhu hello,

Thank you for providing the detailed output from your terminal. It looks like the CoreML export process completed successfully, but you're experiencing issues with missing class labels and preview in the `.mlpackage` file.

To assist you better, could you please confirm the following:
1. Are you using the latest versions of `torch` and `ultralytics`? If not, please upgrade and try the export process again.
2. Could you provide a minimum reproducible example of your code? This will help us understand the context and reproduce the issue on our end. You can refer to our [minimum reproducible example guide](https://docs.ultralytics.com/help/minimum_reproducible_example) for more details.

Regarding your `.pt` file, we cannot accept direct file submissions. However, you can share the relevant code snippets and configurations you used for training and exporting the model.

Here's a quick example of how to export a YOLOv8 model to CoreML and run inference:

```python
from ultralytics import YOLO

# Load the YOLOv8 model
model = YOLO("yolov8n.pt")

# Export the model to CoreML format
model.export(format="coreml")  # creates 'yolov8n.mlpackage'

# Load the exported CoreML model
coreml_model = YOLO("yolov8n.mlpackage")

# Run inference
results = coreml_model("https://ultralytics.com/images/bus.jpg")
```

For more detailed instructions, you can visit our [CoreML integration guide](https://docs.ultralytics.com/integrations/coreml/).

Looking forward to your response so we can help resolve this issue.

**Comment by Sebastianxhj** [comment]
I used the example python code you guys gave. Why does output output confidence and coordination after converting my target detection model to mlpackage format. is it normal to not return processed images? Or is there something wrong with my export process.

**Reply by glenn-jocher** [reply]
@taeja7 the CoreML export returns raw detection outputs (confidence, coordinates) rather than processed images by design. This is normal behavior, as post-processing like bounding box rendering is typically handled within the application layer after inference. For implementation guidance, see the CoreML deployment documentation on handling model outputs in Apple apps.

**Comment by Afool4U** [comment]
Hello! I have trouble when I tried to export yolo11n.pt to coreml model with tracking.
for example, the exist tracking method is like this:
```
# use YOLO tracking
results = yolo_model.track(frame, persist=True, verbose=False)
```
And I've export pt format and reloaded the new coreml format model like this:
```
model.export(format='coreml', nms=True, half=False)
model = ct.models.MLModel('yolo11n.mlpackage')
```
But when I tried to make a tracking on my camera stream, it seems that object 'model' have no attribute named 'track'.
So how could I use tracking by a coreml format model? It's very important to me because the coreml model works well on my macbook.
